File: 13-stack/02-balanced-parenthesis/01-balanced-parenthesis.py
def balanced_parenthesis(parenthesis):
  st = []
  dic = {'}': '{', ')': '(', ']': '['}
  for p in parenthesis:
    if p in dic:
      if not st:
        return False
      if dic[p] != st.pop():
        return False
      continue
    st.append(p)
  return st == []


print(balanced_parenthesis('([])'))
print(balanced_parenthesis('((())'))
print(balanced_parenthesis('([)]'))
print(balanced_parenthesis('{}([()])'))
print(balanced_parenthesis('(()))'))

File: 13-stack/02-balanced-parenthesis/01-balanced-parenthesis.rb
def balanced_parenthesis(parenthesis)
  st = []
  dic = {'}' => '{', ')' => '(', ']' => '['}
  parenthesis.chars.each do |p|
    if dic.keys.include? p
      return false if (st == [] || dic[p] != st.pop)
    elsif
      st.push(p)
    end
  end
  return st == []
end

puts balanced_parenthesis("([])")
puts balanced_parenthesis('((())')
puts balanced_parenthesis('([)]')
puts balanced_parenthesis('{}([()])')
puts balanced_parenthesis('(()))')